src/OrchardCore.Modules/OrchardCore.Contents/AdminMenu.cs

```diff
@@ -55,6 +59,7 @@ public async Task BuildNavigationAsync(string name, NavigationBuilder builder)
                         var ci = await _contentManager.NewAsync(contentTypeDefinition.Name);
                         var cim = await _contentManager.PopulateAspectAsync<ContentItemMetadata>(ci);
                         var createRouteValues = cim.CreateRouteValues;
+                        createRouteValues.Add("returnurl", _httpContextAccessor.HttpContext.Request.PathBase + "/Admin/Contents/ContentItems");
```

Comment:
Here the singularity is to build a route value that is a route itself, we would need to build an `IUrlHelper` with an `IUrlHelperFactory` and a `ViewContext / ActionContext` and then use `.RouteUrl()` with some route values, as it is done through `NavigationManager.BuildMenuAsync()`. This is menu shapes that call `.BuildMenuAsync()` with a `ViewContext / ActionContext` param, but it is not passed when calling all `BuildNavigationAsync()`.

But there is an `IActionContextAccessor` and we have a `ViewContextAccessor` service, so by injecting an `IUrlHelperFactory` we could build an `IUrlHelper`.

OR use the `LinkGenerator` service, there are some methods that need additional params, with some methods you just need to also pass the current `HtpContext` from which it retrieves what it needs, see how it is used in `ContentTypesAdminNodeNavigationBuilder`, e.g. with `_linkGenerator.GetPathByRouteValues()`.

Reply:
done